tests/pauli_rotations/generate_tests_circuit.py

```diff
@@ -0,0 +1,200 @@
+    c1.add_pauli_block(PauliRotation.from_list([Z, X, I, Y], Fraction(1, 8)))
+    c1.add_pauli_block(PauliRotation.from_list([Z, X, I, Y], Fraction(-1, 8)))
+
+    c2 = PauliOpCircuit(7)
+    c2.add_pauli_block(PauliRotation.from_list([X, Y, Z, I, X, Z, Y], Fraction(-1, 4)))
+    c2.add_pauli_block(PauliRotation.from_list([I, Z, Z, I, X, X, X], Fraction(1, 8)))
+    c2.add_pauli_block(Measurement.from_list([Z, X, I, Z, I, I, I]))
+    c2.add_pauli_block(Measurement.from_list([I, I, Z, I, X, X, Z]))
+
+    c3 = PauliOpCircuit(2)
+    c3.add_pauli_block(PauliRotation.from_list([Z, Z], Fraction(1, 4)))
+    c3.add_pauli_block(Measurement.from_list([X, X]))
+
+    return [
+        (c1.circuit_has_measurements(), False),
+        (c2.circuit_has_measurements(), True),
+        (c3.circuit_has_measurements(), True),
+    ]
+
+
+def generate_tests_are_commuting() -> List[Tuple[bool, bool]]:
+    tests_list = list()
+    block1 = PauliRotation.from_list([X, I, Z, I], Fraction(1, 4))
+    block2 = PauliRotation.from_list([I, Z, I, X], Fraction(-1, 4))
+    block3 = PauliRotation.from_list([Z, X, I, Z], Fraction(1, 8))
+    tests_list.append((PauliOpCircuit.are_commuting(block1, block1), True))
+    tests_list.append((PauliOpCircuit.are_commuting(block1, block2), True))
+    tests_list.append((PauliOpCircuit.are_commuting(block1, block3), False))
+    tests_list.append((PauliOpCircuit.are_commuting(block2, block3), True))
+
+    return tests_list
+
+
+def generate_tests_apply_transformation():
+    """
+    Three Litinski Rules
+    1. Test 1 --> P, P` commute - Litinski 4a
+        Note: Whenever two PauliRotation block are checked from commutation, they always
+        commute in the implemented test case
+    2. Test 2 --> P, P` anti commute - Litinski 4a
+        Note: Whenever two PauliRotation block are checked from commutation, they always
+        anti-commute in the implemented test case
+    3. Test 3 --> P, P' commute - Litinski 4b
+    4. Test 4 --> P,P' anti commute - Litinski 4b
+    5. Test 5 --> Controlled Operations, commute - Litinski 4c
+        Tests for this is implicitly included in tests 1 and 2, as controlled operations
+        are converted Pauli Rotations in load_*() methods
+    6. Test 6 --> Controlled Operations, anticommute - Litinski 4c
+        Tests for this is implicitly included in tests 1 and 2, as controlled operations
+        are converted Pauli Rotations in load_*() methods
+
+    `PauliOpCircuit.commute_pi_over_four_rotations()` are also implicitly tested through these tests
+    """
+    tests_list = list()
+    # Test 1:
+    input = PauliOpCircuit(4)
+    input.add_pauli_block(PauliRotation.from_list([X, I, Z, I], Fraction(1, 4)))
+    input.add_pauli_block(PauliRotation.from_list([I, Z, I, X], Fraction(-1, 8)))
+    input.add_pauli_block(PauliRotation.from_list([X, X, I, I], Fraction(-1, 4)))
+    input.add_pauli_block(PauliRotation.from_list([I, I, Z, X], Fraction(1, 8)))
+    input.apply_transformation()
```

Comment:
I don't think it is a good idea to call the function that is being tested inside here. I suggest passing that `input` circuit as an input to `pytest.parametrize`, and inside the test call:

```
assert input.apply_transformation == expected
```

Reply:
sorry - `apply_transformation()` does not return anything so it should be something like this:
```
input.apply_transformation()
assert input == expected
```